pumpkin-py/src/model.rs

```diff
@@ -0,0 +1,185 @@
+        });
+    }
+
+    #[pyo3(signature = (proof=None))]
+    fn satisfy(&self, proof: Option<PathBuf>) -> SatisfactionResult {
+        let proof_log = proof
+            .map(|path| ProofLog::cp(&path, Format::Text, true))
+            .transpose()
+            .map(|proof| proof.unwrap_or_default())
+            .expect("failed to create proof file");
+
+        let options = SolverOptions {
+            proof_log,
+            ..Default::default()
+        };
+
+        let mut solver = Solver::with_options(LearningOptions::default(), options);
+        let variable_map = self.create_variable_map(&mut solver);
+
+        if self.post_constraints(&mut solver, &variable_map).is_err() {
+            return SatisfactionResult::Unsatisfiable();
+        }
+
+        let mut brancher = solver.default_brancher_over_all_propositional_variables();
+
+        match solver.satisfy(&mut brancher, &mut Indefinite) {
+            pumpkin_lib::results::SatisfactionResult::Satisfiable(solution) => {
+                SatisfactionResult::Satisfiable(Solution {
+                    solver_solution: solution,
+                    variable_map,
+                })
+            }
+            pumpkin_lib::results::SatisfactionResult::Unsatisfiable => {
+                SatisfactionResult::Unsatisfiable()
+            }
+            pumpkin_lib::results::SatisfactionResult::Unknown => SatisfactionResult::Unknown(),
+        }
+    }
+}
+
+impl Model {
+    fn create_variable_map(&self, solver: &mut Solver) -> VariableMap {
+        let mut map = VariableMap::default();
+
+        for ModelIntVar {
+            lower_bound,
+            upper_bound,
+            name,
+        } in self.integer_variables.iter()
+        {
+            let _ = map.integers.push(if let Some(name) = name {
+                solver
+                    .new_named_bounded_integer(*lower_bound, *upper_bound, name)
+                    .into()
+            } else {
+                solver
+                    .new_bounded_integer(*lower_bound, *upper_bound)
+                    .into()
+            });
+        }
+
+        map
+    }
+
+    fn post_constraints(
+        &self,
+        solver: &mut Solver,
+        variable_map: &VariableMap,
+    ) -> Result<(), ConstraintOperationError> {
+        for cons in self.constraints.iter() {
+            let ModelConstraint {
+                constraint,
+                premise,
+                tag,
+            } = cons.clone();
+
+            if let Some(premise) = premise {
+                constraint.implied_by(
+                    solver,
+                    variable_map.get_boolean(premise),
+                    tag,
+                    variable_map,
+                )?;
+            } else {
+                constraint.post(solver, tag, variable_map)?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Clone)]
+struct ModelConstraint {
+    constraint: Constraint,
+    premise: Option<BoolVariable>,
```

Comment:
Why the difference in naming? Couldn't we call it a reification literal?

If we want to support both half-reification and full reification in the future then wouldn't it be better to use an enum here?